aptos-move/framework/aptos-framework/sources/Token.move

```diff
@@ -0,0 +1,295 @@
+        let collections = &mut borrow_global_mut<Collections<TokenType>>(account_addr).collections;
+        let gallery = &mut borrow_global_mut<Gallery<TokenType>>(account_addr).gallery;
+
+        let some_data = Option::some(TokenData {
+            description,
+            metadata,
+            name,
+            supply,
+            uri,
+        });
+
+        let (collection_data, gallery_data) = if (supply == 1) {
+            (Option::none(), some_data)
+        } else {
+            (some_data, Option::none())
+        };
+
+        let collection_token = TokenMetadata {
+            id: GUID::id(&GUID::create(account)),
+            data: collection_data,
+        };
+
+        let token_id  = *&collection_token.id;
+        let collection = Table::borrow_mut(collections, &collection_id);
+        if (supply == 1) {
+            Table::insert(&mut collection.claimed_tokens, *&collection_token.id, account_addr)
+        };
+        Table::insert(&mut collection.tokens, *&collection_token.id, collection_token);
+
+        let gallery_token = Token {
+            id: *&token_id,
+            collection: collection_id,
+            balance: supply,
+            data: gallery_data,
+        };
+
+        Table::insert(gallery, *&gallery_token.id, gallery_token);
+        token_id
+    }
+
+    public fun withdraw_token<TokenType: copy + drop + store>(
+        account: &signer,
+        token_id: &ID,
+        amount: u64,
+    ): Token<TokenType> acquires Gallery {
+        let account_addr = Signer::address_of(account);
+
+        let gallery = &mut borrow_global_mut<Gallery<TokenType>>(account_addr).gallery;
+        let balance = Table::borrow(gallery, token_id).balance;
+        assert!(balance >= amount, EINSUFFICIENT_BALANCE);
+
+        if (balance == amount) {
+            let (_key, value) = Table::remove(gallery, token_id);
+            value
+        } else {
+            let token = Table::borrow_mut(gallery, token_id);
+            token.balance = balance - amount;
+            Token {
+                id: *&token.id,
+                collection: *&token.collection,
+                balance: amount,
+                data: *&token.data,
+            }
+        }
+    }
+
+    public fun deposit_token<TokenType: copy + drop + store>(
+        account: &signer,
+        token: Token<TokenType>,
+    ) acquires Collections, Gallery {
+        let account_addr = Signer::address_of(account);
+        if (!exists<Gallery<TokenType>>(account_addr)) {
+            initialize_gallery<TokenType>(account)
+        };
+
+        let creator_addr = GUID::id_creator_address(&token.collection);
+        let collections = &mut borrow_global_mut<Collections<TokenType>>(creator_addr).collections;
+        let collection = Table::borrow_mut(collections, &token.collection);
+        if (Option::is_some(&token.data) && Option::borrow(&token.data).supply == 1) {
+          Table::remove(&mut collection.claimed_tokens, &token.id);
```

Comment:
this should be done in withdraw?

Reply:
need to move to a generalized function as discussed today so that non-galleries can make this call

Reply:
Per the conversation today, the behavior of claimed tokens is ambiguous -- my intent was to capture the last known holder, and withdraw still implies it is owned by the current user just not in a gallery, so this is technically correct. Will add documentation on semantics here and also expose this api to be called by passing in the token and getting it back -- e.g. proof of ownership.